keras_cv/metrics/coco/recall.py

```diff
@@ -12,11 +12,173 @@
+
+    Usage:
+        COCORecall accepts two Tensors as input to it's `update_state` method.
+        These Tensors represent bounding boxes in `corners` format.  Utilities
+        to convert Tensors from `xywh` to `corners` format can be found in
+        `keras_cv.utils.bbox`.
+
+        Each image in a dataset may have a different number of bounding boxes,
+        both in the ground truth dataset and the prediction set.  In order to
+        account for this, you may either pass a `tf.RaggedTensor`, or pad Tensors
+        with `-1`s to indicate unused boxes.  A utility function to perform this
+        padding is available at `keras_cv_.utils.bbox.pad_bbox_batch_to_shape`.
+
+        ```
+        coco_recall = COCORecall(
+            max_detections=100,
+            category_ids=[1],
+            area_range=(32**2, 64**2),
+        )
+
+        y_true = np.array([[[0, 0, 10, 10, 1], [5, 5, 10, 10, 1]]]).astype(np.float32)
+        y_pred = np.array([[[0, 0, 10, 10, 1, 1.0], [5, 5, 10, 10, 1, 0.9]]]).astype(
+            np.float32
+        )
+        coco_recall.update_state(y_true, y_pred)
+        coco_recall.result()
+        # > 0.0
+        ```
+    """
+
+    def __init__(
+        self,
+        category_ids,
+        iou_thresholds=None,
+        area_range=(0, 1e9**2),
+        max_detections=100,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        # Initialize parameter values
+
+        iou_thresholds = iou_thresholds or [x / 100.0 for x in range(50, 100, 5)]
+        self.iou_thresholds = iou_thresholds
+        self.category_ids = category_ids
+
+        self.area_range = area_range
+        self.max_detections = max_detections
+
+        # Initialize result counters
+        num_thresholds = len(iou_thresholds)
+        num_categories = len(category_ids)
+
+        self.true_positives = self.add_weight(
+            name="true_positives",
+            shape=(num_thresholds, num_categories),
+            dtype=tf.float32,
+            initializer=initializers.Zeros(),
+        )
+        self.ground_truth_boxes = self.add_weight(
+            name="ground_truth_boxes",
+            shape=(num_categories,),
+            dtype=tf.float32,
+            initializer=initializers.Zeros(),
+        )
+
+    def reset_state(self):
+        self.true_positives.assign(tf.zeros_like(self.true_positives))
+        self.ground_truth_boxes.assign(tf.zeros_like(self.ground_truth_boxes))
+
+    @tf.function
+    def update_state(self, y_true, y_pred, sample_weight=None):
+        """
+        Args:
+            y_true: a bounding box Tensor in corners format.
+            y_pred: a bounding box Tensor in corners format.
+            sample_weight: Currently unsupported.
+        """
+        if sample_weight:
+            raise NotImplementedError(
+                "sample_weight is not yet supported in keras_cv COCO metrics."
+            )
+
+        if isinstance(y_true, tf.RaggedTensor):
```

Comment:
We might want to add a TODO here for native ragged input support, since convert ragged tensor to dense tensor could potentially lead to memory issue and cause OOM.